hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/utils/AzcopyToolHelper.java

```diff
@@ -176,70 +178,115 @@ private File findHadoopAzureDir(File dir) {
+      extractProcess.waitFor();
+
+      // Rename the azcopy_linux_amd64_* directory to 'azcopy' and move it to the hadoop-azure directory
+      String renameCmd = MOVE_CMD + hadoopAzureDir.getAbsolutePath() + AZCOPY_DOWNLOADED_DIR_NAME + azcopyDirPath;
+      String[] renameCmdArr = {AZCOPY_CMD_SHELL, AZCOPY_CMD_OPTION, renameCmd};
+      Process renameProcess = Runtime.getRuntime().exec(renameCmdArr);
+      renameProcess.waitFor();
+
+      // Remove the downloaded tarball and azcopy folder
+      String cleanupCmd = REMOVE_CMD + hadoopAzureDir.getAbsolutePath() + AZCOPY_DOWNLOADED_TAR_NAME;
+      String[] cleanupCmdArr = {AZCOPY_CMD_SHELL, AZCOPY_CMD_OPTION, cleanupCmd};
+      Process cleanupProcess = Runtime.getRuntime().exec(cleanupCmdArr);
+      cleanupProcess.waitFor();
+
+      // Set the execute permission on the azcopy executable
+      String chmodCmd = CHMOD_CMD + azcopyDirPath;
+      String[] chmodCmdArr = {AZCOPY_CMD_SHELL, AZCOPY_CMD_OPTION, chmodCmd};
+      Process chmodProcess = Runtime.getRuntime().exec(chmodCmdArr);
+      chmodProcess.waitFor();
     }
-    String cleanupCmd = REMOVE_CMD + path;
-    String[] cleanupCmdArr = {AZCOPY_CMD_SHELL, AZCOPY_CMD_OPTION, cleanupCmd};
-    Process cleanupProcess = Runtime.getRuntime().exec(cleanupCmdArr);
-    cleanupProcess.waitFor();
   }
 
-  private static void createFileCreationScript(String folderPath, String scriptName, String azcopyPath, String sasToken, String containerName) {
-    String blobPath = containerName + "?" + sasToken; // construct the blob path
-    String scriptContent = "blobPath=\"" + blobPath + "\"\n"
-        + "echo $blobPath\n"
-        + azcopyPath + "/azcopy copy \"" + azcopyPath + "/NOTICE.txt\" $blobPath\n"; // construct the script content
-    File scriptFile = new File(folderPath, scriptName);
+  private boolean fileExists(String filePath) {
+    File file = new File(filePath);
+    return file.exists();
+  }
+
+  /**
+   * Create a shell script with exclusive file lock.
+   * Different JVMs can try to create this file in case of CI runs. To avoid conflicts,
+   * We are using exclusive file lock so that a single script exists and used by all the JVMs.
+   * @param scriptPath to be created
+   * @param scriptContent to be written in the script.
+   */
+  private void createShellScript(String scriptPath, String scriptContent) {
+    RandomAccessFile file = null;
+    FileLock fileLock = null;
     try {
-      FileWriter writer = new FileWriter(scriptFile);
-      writer.write(scriptContent);
-      writer.close();
-      boolean written = scriptFile.setExecutable(true); // make the script executable
-      System.out.println("Script created at " + scriptFile.getAbsolutePath());
+      file = new RandomAccessFile(scriptPath, "rw");
+      FileChannel fileChannel = file.getChannel();
+
+      // Try acquiring the lock
+      fileLock = fileChannel.tryLock();
+      if (fileLock != null) {
```

Comment:
Lets check once more that file is not existing.

Reason being, a process could be line 246, while the other process is taking the lock and closing it. And after that context-switch happen in which first process comes to line 249, now it would create again or make the data inconsistent.